x-pack/plugins/infra/public/utils/remote_state/remote_graphql_state.ts

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ApolloError, ApolloQueryResult } from 'apollo-client';
+import { DocumentNode } from 'graphql';
+import { Action as ReduxAction } from 'redux';
+import { Epic } from 'redux-observable';
+import { from, merge, Observable, pipe } from 'rxjs';
+import { catchError, filter, map, startWith, switchMap, withLatestFrom } from 'rxjs/operators';
+import { Action, ActionCreator, actionCreatorFactory, Failure, Success } from 'typescript-fsa';
+import { reducerWithInitialState } from 'typescript-fsa-reducers';
+
+import { createSelector } from 'reselect';
+import { InfraApolloClient } from '../../lib/lib';
+import {
+  isFailureLoadingResult,
+  isIdleLoadingProgress,
+  isRunningLoadingProgress,
+  isSuccessLoadingResult,
+  isUninitializedLoadingResult,
+  LoadingPolicy,
+  LoadingProgress,
+  LoadingResult,
+} from '../loading_state';
+
+export interface GraphqlState<State> {
+  current: LoadingProgress<OperationInfo<any>>;
+  last: LoadingResult<OperationInfo<any>>;
+  data: State | undefined;
+}
+
+interface OperationInfo<Variables> {
+  operationKey: string;
+  variables: Variables;
+}
+
+type ResolveDonePayload<Variables, Data> = Success<Variables, ApolloQueryResult<Data>>;
+type ResolveFailedPayload<Variables, Error> = Failure<Variables, Error>;
+
+interface OperationActionCreators<Data, Variables, Error = ApolloError> {
+  resolve: ActionCreator<Variables>;
+  resolveStarted: ActionCreator<Variables>;
+  resolveDone: ActionCreator<ResolveDonePayload<Variables, Data>>;
+  resolveFailed: ActionCreator<ResolveFailedPayload<Variables, Error>>;
+}
+
+export const createGraphqlInitialState = <State>(initialData?: State): GraphqlState<State> => ({
+  current: {
+    progress: 'idle',
```

Comment:
If this is loading state, could it be named something with `loading`? `progress` makes me think of a 0-100% range or something similar and I believe we don't have that.

Reply:
yes, I will try to streamline this a bit

Reply:
Sure!